File: src/panda_simulation/panda_control_MAIF/include/AIC.py
#!/usr/bin/env python2.7
import rospy
from std_msgs import Float64MultiArray, MultiArrayDimension, Float64
from geometry_msgs import PoseStamped 
from sensor_msgs import JointState, Image 
import numpy as np
import scipy.io
import math
import sys
#include <vector>
#include <cmath>
#include <eigen3/Eigen/Core>
#include <eigen3/Eigen/Dense>
#include <iostream>
#include <stdlib.h>



class AIC_agent:

    # Number of joints used
    N_JOINTS = 7

    # Image parameters
    IMG_WIDTH = 256
    IMG_HEIGHT = 256

  def __init__(self, Robot):
        
        self.jointStatesCallback(self,msg);
        self.initVariables(self);
        self.minimiseF_Prop(self);
        self.dF_callback(self, dF_dmu_vis);
        self.dF_callback_attr(self, dF_dmu_attr_p);
        self.minimiseF_PixelAI(self);
        self.computeActions(self);
        self.dataReady(self);
        self.setGoal(self, desiredPos);
        self.getSPE(self);


        self.SigmaP_yq0 = np.zeros((self.N_JOINTS,self.N_JOINTS)
        self.SigmaP_yq1 = np.zeros((self.N_JOINTS,self.N_JOINTS)
        self.SigmaP_mu = np.zeros((self.N_JOINTS,self.N_JOINTS)
        self.SigmaP_muprime = np.zeros((self.N_JOINTS,self.N_JOINTS)
        self.mu = np.zeros((1,self.N_JOINTS)
        self.mu_p = np.zeros((1,self.N_JOINTS)
        self.mu_pp = np.zeros((1,self.N_JOINTS)
        self.mu_dot = np.zeros((1,self.N_JOINTS)
        self.mu_dot_p = np.zeros((1,self.N_JOINTS)
        self.mu_dot_pp = np.zeros((1,self.N_JOINTS)
        self.jointPos = np.zeros((1,self.N_JOINTS)
        self.jointVel = np.zeros((1,self.N_JOINTS)
        self.mu_d = np.zeros((1,self.N_JOINTS)
        self.a = np.zeros((1,self.N_JOINTS)
        self.a_dot = np.zeros((1,self.N_JOINTS)
        self.s_p = np.zeros((1,self.N_JOINTS)


        self.AIC_mu = FloatMultiArray()
        self.AIC_mu_p = FloatMultiArray()
        self.AIC_mu_pp = FloatMultiArray()
        self.SPE = FloatMultiArray()
        self.dF_dg_mu_vis = FloatMultiArray()
        self.dF_dmu_attr = FloatMultiArray()


        self.datareceived = 0  

        self.var_mu = 5.0
        self.var_muprime = 10.0  
        self.var_q = 1
        self.var_qdot = 1

        # Learning rates for the gradient descent (found that a ratio of 60 works good)
        self.k_mu = 11.67
        self.k_a = 700
        #Attractor gain
        self.beta = 1

        for i in range(N_JOINTS):
            SigmaP_yq0[i, i] = 1/var_q
            SigmaP_yq1[i, i] = 1/var_qdot
            SigmaP_mu[i, i] = 1/var_mu
            SigmaP_muprime[i, i] = 1/var_muprime

        self.h = 0.001
        
        #Publisher
        self.tauPub1 = rospy.Publisher("/robot1/panda_joint1_controller/command", Float64, queue_size = 20)
        self.tauPub2 = rospy.Publisher("/robot1/panda_joint2_controller/command", Float64, queue_size = 20)
        self.tauPub3 = rospy.Publisher("/robot1/panda_joint3_controller/command", Float64, queue_size = 20)
        self.tauPub4 = rospy.Publisher("/robot1/panda_joint4_controller/command", Float64, queue_size = 20)
        self.tauPub5 = rospy.Publisher("/robot1/panda_joint5_controller/command", Float64, queue_size = 20)
        self.tauPub6 = rospy.Publisher("/robot1/panda_joint6_controller/command", Float64, queue_size = 20)
        self.tauPub7 = rospy.Publisher("/robot1/panda_joint7_controller/command", Float64, queue_size = 20)

        self.IFE_pub = rospy.Publisher("panda_free_energy", Float64, queue_size = 10)
        self.SPE_pub = rospy.Publisher("panda_SPE", Float64, queue_size = 10)
        

        self.beliefs_mu_pub = rospy.Publisher("beliefs_mu",Float64,queue_size = 20)
        self.beliefs_mu_p_pub = rospy.Publisher("beliefs_mu_p",Float64,queue_size = 20)
        self.beliefs_mu_pp_pub = rospy.Publisher("beliefs_mu_pp",Float64,queue_size = 20)

        #Subscribers
        sensorSub = rospy.Subscriber("/robot1/joint_states", JointState, self.jointStatesCallback);
        #sub = rospy.Subscriber("/perception/pub_dF_vis",, dF_callback);
        #sub_ = rospy.Subscriber("perception/pub_dF_attr",1, AIC::dF_callback_attr);



  def jointStatesCallback(self, msg):
      
        for i in range(N_JOINTS):
        
            self.jointPos[i] = msg.position[i];
            self.jointVel[i] = msg.velocity[i];

        if (self.dataReceived == 0):

           self.dataReceived = 1
           self.mu = self.jointpos
           self.mu_p = self.jointVel;


  def minimiseF_prop(self):
    
        # Compute single sensory prediction errors
        self.SPEq = (np.transpose(self.jointPos)-np.transpose(self.mu))*self.SigmaP_yq0*(self.jointPos-self.mu);
        self.SPEdq = (np.transpose(self.jointVel)-np.transpose(self.mu_p))*self.SigmaP_yq1*(self.jointVel-self.mu_p);
        self.SPEmu_p = (np.transpose(self.mu_p)+np.transpose(self.mu)-np.transpose(self.mu_d))*self.SigmaP_mu*(self.mu_p+self.mu-self.mu_d);
        self.SPEmu_pp = (np.transpose(self.mu_pp)+np.transpose(self.mu_p))*self.SigmaP_muprime*(self.mu_pp+self.mu_p);
        
        # Free-energy as a sum of squared values (i.e. sum the SPE)
        self.F_Prop.data = self.SPEq + self.SPEdq + self.SPEmu_p + self.SPEmu_pp;


       #Free-energy minimization using gradient descent and beliefs update
       self.mu_dot = self.mu_p - self.k_mu*(-self.SigmaP_yq0*(self.jointPos-self.mu)+self.SigmaP_mu*(self.mu_p+self.mu-self.mu_d));
   
       self.mu_dot_p = self.mu_pp - self.k_mu*(-self.SigmaP_yq1*(self.jointVel-self.mu_p)+self.SigmaP_mu*(self.mu_p+self.mu-self.mu_d)+self.SigmaP_muprime*(self.mu_pp+self.mu_p));
       self.mu_dot_pp = - self.k_mu*(self.SigmaP_muprime*(self.mu_pp+self.mu_p));

       #Belifs update
       self.mu = self.mu + self.h*self.mu_dot;             # Belief about the position
       self.mu_p = self.mu_p + self.h*self.mu_dot_p;       # Belief about motion of mu
       self.mu_pp = self.mu_pp + self.h*self.mu_dot_pp;    # Belief about motion of mu'

       for i in range(N_JOINTS):
            self.AIC_mu.data[i] = self.mu[i];
            self.AIC_mu_p.data[i] = self.mu_p[i];
            self.AIC_mu_pp.data[i] = self.mu_pp[i];

       self.SPE.data[0] = self.SPEq;
       self.SPE.data[1] = self.SPEdq;


       self.computeActions()


       self.IFE_pub.publish(self.F_Prop)

       # Sensory prediction error publisher
       self.SPE_pub.publish(self.SPE)

       # Publish beliefs
       self.beliefs_mu_pub.publish(self.AIC_mu)
       self.beliefs_mu_p_pub.publish(self.AIC_mu_p)
       self.beliefs_mu_pp_pub.publish(self.AIC_mu_pp)


   def computeActions(self):

      self.a = self.a-self.h*self.k_a*(self.SigmaP_yq1*(self.jointVel-self.mu_p)+self.SigmaP_yq0*(self.jointPos-self.mu));

      # Set the toques from u and publish
      self.tau1.data = self.a(0) 
      self.tau2.data = self.a(1) 
      self.tau3.data = self.a(2) 
      self.tau4.data = self.a(3)
      self.tau5.data = self.a(4)
      self.tau6.data = self.a(5)
      self.tau7.data = self.a(6)
      # Publishing
      self.tauPub1.publish(self.tau1)
      self.tauPub2.publish(self.tau2)
      self.tauPub3.publish(self.tau3)
      self.tauPub4.publish(self.tau4)
      self.tauPub5.publish(self.tau5)
      self.tauPub6.publish(self.tau6)
      self.tauPub7.publish(self.tau7)


   def dataReady(self):

       if (dataReceived ==1):
            return 1
       else:
            return 0


   def setGoal(self, desiredPos):

       for i in range(N_JOINTS):
           self.mu_d[i] = desiredPos[i]

   

       

      
     



           

       
      

        
     



 
 
 
 
        

        

File: src/panda_simulation/panda_control_PAIF/src/classes/AIC.cpp
/*
 * File:   AIC.cpp
 * Author: Corrado Pezzato, TU Delft, DCSC
 *
 * Created on April 14th, 2019
 *
 * Class to perform active inference control of the 7DOF Franka Emika Panda robot.
 * Definition of the methods contained in AIC.h
 *
 */

#include "AIC.h"

  // Constructor which takes as argument the publishers and initialises the private ones in the class
  AIC::AIC(int whichRobot){

      // Initialize publishers on the topics /robot1/panda_joint*_controller/command for the joint efforts
      tauPub1 = nh.advertise<std_msgs::Float64>("/robot1/panda_joint1_controller/command", 20);
      tauPub2 = nh.advertise<std_msgs::Float64>("/robot1/panda_joint2_controller/command", 20);
      tauPub3 = nh.advertise<std_msgs::Float64>("/robot1/panda_joint3_controller/command", 20);
      tauPub4 = nh.advertise<std_msgs::Float64>("/robot1/panda_joint4_controller/command", 20);
      tauPub5 = nh.advertise<std_msgs::Float64>("/robot1/panda_joint5_controller/command", 20);
      tauPub6 = nh.advertise<std_msgs::Float64>("/robot1/panda_joint6_controller/command", 20);
      tauPub7 = nh.advertise<std_msgs::Float64>("/robot1/panda_joint7_controller/command", 20);
      sensorSub = nh.subscribe("/robot1/joint_states", 1, &AIC::jointStatesCallback, this);
      end_effSub = nh.subscribe("/gazebo/link_states", 1, &AIC::end_effStatesCallback, this);
      // Publisher for the free-energy and sensory prediction errors
      IFE_pub = nh.advertise<std_msgs::Float64>("panda_free_energy", 10);
      SPE_pub = nh.advertise<std_msgs::Float64MultiArray>("panda_SPE", 10);

      // Publishers for beliefs
      beliefs_mu_pub = nh.advertise<std_msgs::Float64MultiArray>("beliefs_mu", 10);
      beliefs_mu_p_pub = nh.advertise<std_msgs::Float64MultiArray>("beliefs_mu_p", 10);
      beliefs_mu_pp_pub = nh.advertise<std_msgs::Float64MultiArray>("beliefs_mu_pp", 10);

      // Publishers for beliefs
      beliefs_mu_pub = nh.advertise<std_msgs::Float64MultiArray>("beliefs_mu", 10);
      beliefs_mu_p_pub = nh.advertise<std_msgs::Float64MultiArray>("beliefs_mu_p", 10);
      beliefs_mu_pp_pub = nh.advertise<std_msgs::Float64MultiArray>("beliefs_mu_pp", 10);
       
    // Initialize the variables for thr AIC
    AIC::initVariables();
  }
  AIC::~AIC(){}

  void   AIC::jointStatesCallback(const sensor_msgs::JointState::ConstPtr& msg)
  {
    // Save joint values
    for( int i = 0; i < 7; i++ ) {
      jointPos(i) = msg->position[i];
      //jointPos(i) = jointPos(i) + q_noise(i);
      jointVel(i) = msg->velocity[i];
    }
    // If this is the first time we read the joint states then we set the current beliefs
    if (dataReceived == 0){
      // Track the fact that the encoders published
      dataReceived = 1;
      // The first time we retrieve the position we define the initial beliefs about the states
      mu = jointPos;
      mu_p = jointVel;
    }
  }


void   AIC::end_effStatesCallback(const gazebo_msgs::LinkStates::ConstPtr& msg)
  {
    // Save joint values
    end_eff_x = msg->pose[9].position.x;
    end_eff_y = msg->pose[9].position.y;
    end_eff_z = msg->pose[9].position.z;

    end_eff = sqrt(pow(end_eff_x,2) + pow(end_eff_y,2) + pow(end_eff_z,2));

  }

  void   AIC::initVariables(){

    // Support variable
    dataReceived = 0;

    // Variances associated with the beliefs and the sensory inputs
    var_mu = 5.0;
    var_muprime = 10.0;
    var_q = 1;
    var_qdot = 1;

    // Learning rates for the gradient descent (found that a ratio of 60 works good)
    k_mu = 11.67;
    k_a = 700;

    // Precision matrices (first set them to zero then populate the diagonal)
    SigmaP_yq0 = Eigen::Matrix<double, 7, 7>::Zero();
    SigmaP_yq1 = Eigen::Matrix<double, 7, 7>::Zero();
    SigmaP_mu = Eigen::Matrix<double, 7, 7>::Zero();
    SigmaP_muprime = Eigen::Matrix<double, 7, 7>::Zero();

    for( int i = 0; i < SigmaP_yq0.rows(); i = i + 1 ) {
      SigmaP_yq0(i,i) = 1/var_q;
      SigmaP_yq1(i,i) = 1/var_qdot;
      SigmaP_mu(i,i) = 1/var_mu;
      SigmaP_muprime(i,i) = 1/var_muprime;
    }

    // Initialize control actions
    u << 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0;

    // Initialize prior beliefs about the second ordet derivatives of the states of the robot
    mu_pp << 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0;

    end_eff_d << 1.05342339042937, 0.683566227681276, 1.01240873890366, 0.920884331281194;
    end_eff_d_x << 0.114635239379695, 0.328372326019533, 0.559577132421758, 0.379710561368265;
    end_eff_d_y << 0.00209295101959208, 0.511476097394308, 0.003013607548644, -0.591409215177049;
    end_eff_d_z << 1.03170767445082, 0.312485466168603, 0.84239241467989, 0.594437411655472;




    // Integration step
    h = 0.001;

    // Resize Float64MultiArray messages
    AIC_mu.data.resize(7);
    AIC_mu_p.data.resize(7);
    AIC_mu_pp.data.resize(7);
    SPE.data.resize(2);
  }

  void AIC::noise(){

    q_noise(0) = distribution_q0(generator);
    q_noise(1) = distribution_q1(generator);
    q_noise(2) = distribution_q2(generator);
    q_noise(3) = distribution_q3(generator);
    q_noise(4) = distribution_q4(generator);
    q_noise(5) = distribution_q5(generator);
    q_noise(6) = distribution_q6(generator);
    std::ofstream fileWrite("noise.txt", std::ios::app );


  }
  void AIC::minimiseF(int j, int d){


    // Compute single sensory prediction errors
    SPEq = (jointPos.transpose()-mu.transpose())*SigmaP_yq0*(jointPos-mu);
    SPEdq = (jointVel.transpose()-mu_p.transpose())*SigmaP_yq1*(jointVel-mu_p);
    SPEmu_p = (mu_p.transpose()+mu.transpose()-mu_d.transpose())*SigmaP_mu*(mu_p+mu-mu_d);
    SPEmu_pp = (mu_pp.transpose()+mu_p.transpose())*SigmaP_muprime*(mu_pp+mu_p);

    // Free-energy as a sum of squared values (i.e. sum the SPE)
    F.data = SPEq + SPEdq + SPEmu_p + SPEmu_pp;

    // Free-energy minimization using gradient descent and beliefs update
    mu_dot = mu_p - k_mu*(-SigmaP_yq0*(jointPos-mu)+SigmaP_mu*(mu_p+mu-mu_d));
    mu_dot_p = mu_pp - k_mu*(-SigmaP_yq1*(jointVel-mu_p)+SigmaP_mu*(mu_p+mu-mu_d)+SigmaP_muprime*(mu_pp+mu_p));
    mu_dot_pp = - k_mu*(SigmaP_muprime*(mu_pp+mu_p));

    // Belifs update
    mu = mu + h*mu_dot;             // Belief about the position
    mu_p = mu_p + h*mu_dot_p;       // Belief about motion of mu
    mu_pp = mu_pp + h*mu_dot_pp;    // Belief about motion of mu'
    //AIC::noise();
    // Publish beliefs as Float64MultiArray
    for (int i=0;i<7;i++){
       AIC_mu.data[i] = mu(i);
       AIC_mu_p.data[i] = mu_p(i);
       AIC_mu_pp.data[i] = mu_pp(i);
    }
    // Define SPE message
    SPE.data[0] = SPEq;
    SPE.data[1] = SPEdq;

    // Calculate and send control actions
    AIC::computeActions();

    // Publish free-energy
    IFE_pub.publish(F);

    // Sensory prediction error publisher
    SPE_pub.publish(SPE);

    // Publish beliefs
    beliefs_mu_pub.publish(AIC_mu);
    beliefs_mu_p_pub.publish(AIC_mu_p);
    beliefs_mu_pp_pub.publish(AIC_mu_pp);

    //err_end_eff  = abs(end_eff_d_x[d] - end_eff_x)  + abs(end_eff_d_y[d] - end_eff_y)  + abs(end_eff_d_z[d] - end_eff_z);

    //err_end_eff = end_eff - end_eff_d(d);
    //std::ofstream fileWrite("err_end_eff_noise.txt", std::ios::app);
    //fileWrite << err_end_eff;
    //fileWrite << "\n";
    //fileWrite.close(); 
  }

  void   AIC::computeActions(){
    // Compute control actions through gradient descent of F
    u = u-h*k_a*(SigmaP_yq1*(jointVel-mu_p)+SigmaP_yq0*(jointPos-mu));

    // Set the toques from u and publish
    tau1.data = u(0); tau2.data = u(1); tau3.data = u(2); tau4.data = u(3);
    tau5.data = u(4); tau6.data = u(5); tau7.data = u(6);
    // Publishing
    tauPub1.publish(tau1); tauPub2.publish(tau2); tauPub3.publish(tau3);
    tauPub4.publish(tau4); tauPub5.publish(tau5); tauPub6.publish(tau6);
    tauPub7.publish(tau7);
  }

  int AIC::dataReady(){
    // Method to control if the joint states have been received already,
    // used in the main function
    if(dataReceived==1)
      return 1;
    else
      return 0;
  }

  void AIC::setGoal(std::vector<double> desiredPos){
    for(int i=0; i<desiredPos.size(); i++){
      mu_d(i) = desiredPos[i];
    }
  }

  std_msgs::Float64MultiArray  AIC::getSPE(){
    return(SPE);
  }
